### ISSUE TITLE
I have specified LogoutUri but it does not call those urls

### ISSUE THREAD

**Issue by slepoiKot** [IdentityServer/IdentityServer3#2879]
Hi, could you please help me a bit with IdSrv3 + OpenId + SSO.
I have specified LogoutUri for all my clients.

My SignOut method is:
```cs
        [AllowAnonymous]
        [HttpGet("~/signout")]
        [HttpPost("~/signout")]
        public async Task SignOut()
        {
            _logger.LogVerbose($"SignOut()");
            await HttpContext.Authentication.SignOutAsync("Cookies");
            await HttpContext.Authentication.SignOutAsync(OpenIdConnectDefaults.AuthenticationScheme);
        }
```
it makes IdSrv to signout my clients, I see in log:
```
23:29:10 [Information] Protocol: HTTP/1.1  Method: GET  ContentType:   ContentLength:   Scheme: http  Host: idsrv:18942  PathBase:   Path: /logout  QueryString: ?id=8922f435f4c115197b87bb75c94f218c
23:29:10 [Information] Logout prompt for subject: "user1"
23:29:10 [Information] SignOutMessage present (from client "portal") and RequireSignOutPrompt is false, performing logout
23:29:10 [Information] Logout endpoint submitted
23:29:10 [Information] Logout requested for subject: "user1"
23:29:10 [Information] Clearing cookies
23:29:10 [Information] rendering logged out page
23:29:10 [Information] Request finished in 0ms 200 text/html; charset=utf-8
23:29:10 [Information] Protocol: HTTP/1.1  Method: GET  ContentType:   ContentLength:   Scheme: http  Host: idsrv:18942  PathBase:   Path: /connect/endsessioncallback  QueryString: ?sid=665ac7af61e7d269164369e8d4e57651
23:29:10 [Information] End session callback requested
23:29:10 [Debug] Client end session iframe URLs: "http://api:19806/SignoutCleanup?sid=665ac7af61e7d269164369e8d4e57651, http://portal:41868/SignoutCleanup?sid=665ac7af61e7d269164369e8d4e57651"
23:29:10 [Information] Request finished in 0.0047ms 200 text/html; charset=utf-8
```
I have *portal* and *api* clients, but none of them get called.
If I make call to http://api:19806/SignoutCleanup?sid=665ac7af61e7d269164369e8d4e57651 and to http://portal:41868/SignoutCleanup?sid=665ac7af61e7d269164369e8d4e57651 it will successfully signout each of them.

```cs
        [AllowAnonymous]
        [HttpGet("~/SignoutCleanup")]
        [HttpPost("~/SignoutCleanup")]
        public async Task SignoutCleanup(string sid)
        {
            _logger.LogVerbose($"SignoutCleanup({sid})");

            var sidClaim = User.FindFirst("sid");
            _logger.LogVerbose($"'{User.Identity.Name}' IsAuthenticated:[{User.Identity.IsAuthenticated}] sidClaim: {sidClaim}");

            if (sidClaim != null && sidClaim.Value == sid)
            {
                await SignOut();
            }
        }
```

client example:
```cs
new Client
{
    ClientName = "Portal",
    ClientId = "portal",
	Flow = Flows.Hybrid,
	ClientSecrets = new List<Secret>
    {
        new Secret("secret".Sha256())
    },
    RequireConsent = false,
    
    LogoutUri = "http://portal:41868/SignoutCleanup",
    LogoutSessionRequired = true,

    RedirectUris = new List<string>
    {
        "http://portal:41868/signin-oidc",
        "http://portal:41868/",
    },
    PostLogoutRedirectUris = new List<string>
    {
        "http://portal:41868/",
    },
...
```

What do I do wrong? Or did I misunderstand something. Thank you in advance, I've read many times documentation, did google it and found nothing that might help me to get it work.

**Comment by dturska** [comment]
what does your startup class look like for one of your clients? Are you passing the id_token during a logout request?


app.UseOpenIdConnectAuthentication(new OpenIdConnectAuthenticationOptions
            {
                ClientId = Constants.ClientCredentials.ClientId,
                ClientSecret = Constants.ClientCredentials.ClientSecret,
                Authority = Constants.Endpoints.IdentityServerEndpoint,
                SignInAsAuthenticationType = Constants.AuthenticationTypeCookies,
                ResponseType = string.Join(" ", Constants.OpenIdConnectSettings.ResponseTypes),
                Scope = string.Join(" ", Constants.OpenIdConnectSettings.ScopesRequesting),
               
               
                ProtocolValidator = new Microsoft.IdentityModel.Protocols.OpenIdConnectProtocolValidator
                {
                    RequireNonce = true,
                },

                Notifications = new OpenIdConnectAuthenticationNotifications
                {
                    RedirectToIdentityProvider = async context =>
                    {
                        context.ProtocolMessage.RedirectUri = Constants.Endpoints.BaseAddress;

                        **if (context.ProtocolMessage.RequestType == Microsoft.IdentityModel.Protocols.OpenIdConnectRequestType.LogoutRequest)**
                        {
                            var idTokenHint = context.OwinContext.Authentication.User.FindFirst("id_token");

                            if(idTokenHint != null)
                            {
                                context.ProtocolMessage.PostLogoutRedirectUri = Constants.Endpoints.BaseAddress;
                                context.ProtocolMessage.IdTokenHint = idTokenHint.Value;
                            }

                        }
                        await Task.FromResult(context);
                    },

**Comment by slepoiKot** [comment]
Yes, I pass `id_token`, but do not return `context`.
```cs
OnRedirectToAuthenticationEndpoint = context =>
{
    log.LogVerbose("OnRedirectToAuthenticationEndpoint");
    if (context.ProtocolMessage.RequestType == OpenIdConnectRequestType.LogoutRequest)
    {
        var idTokenHint = context.HttpContext.User.FindFirst("id_token");

        if (idTokenHint != null)
        {
            context.ProtocolMessage.IdTokenHint = idTokenHint.Value;
        }
    }

    return Task.FromResult(0);
}
```
Are you sure it's something at client side? IdSrv does not call clients' logout URL, but if I do call that URL, those clients get signed out normally.

Look at IdSrv journal, it looks like it tried to make calls. But they are did not happen at client side.
```
23:29:10 [Debug] Client end session iframe URLs: "http://api:19806/SignoutCleanup?sid=665ac7af61e7d269164369e8d4e57651, http://portal:41868/SignoutCleanup?sid=665ac7af61e7d269164369e8d4e57651"
23:29:10 [Information] Request finished in 0.0047ms 200 text/html; charset=utf-8
```

**Comment by brockallen** [comment]
Use the dev tools in your browser and see if the iframes are rendered and if your browser is calling them. If IdSvr is on https and the client is http then the browser won't call them,.

**Comment by slepoiKot** [comment]
iframes have not been rendered
IdSrv and client both use http
I think I do something wrong. But can't find it 8(

**Comment by brockallen** [comment]
Any update?

**Comment by slepoiKot** [comment]
I'm stuck with with single sign out. I see no iframes rendered. And don't know where to drill further. May be I misunderstood official manual regarding http sign-out here: [https://identityserver.github.io/Documentation/docsv2/advanced/signout-http.html](https://identityserver.github.io/Documentation/docsv2/advanced/signout-http.html)

**Comment by brockallen** [comment]
> I see no iframes rendered.

Look at the DOM, not the rendered HTML.

**Comment by brockallen** [comment]
All set? Can we close?

**Comment by slepoiKot** [comment]
I had no time to try. Anyway, thank you for assistance. I'll get back to this in a while.

**Comment by brockallen** [comment]
The only other thing I can suggest is for you to run the sample (MVC Hybrid) and see how it handles SLO.

**Comment by josuemb** [comment]
Hint: I am facing the same issue. In my specific case I discovered that using empty Claim values causes the same symptom. So I modified my code to generate claims only in case the value is not empty.

Example:

            List<Claim> userClaims = new List<Claim>();
            if (!string.IsNullOrWhiteSpace(user[FIELD_USUARIO_CVE]))
            {
                userClaims.Add(new Claim(Constants.ClaimTypes.Subject, user[FIELD_USUARIO_CVE].Trim()));
            }
            if (!string.IsNullOrWhiteSpace(user[FIELD_USUARIO_NOMBRE].Trim()))
            {
                userClaims.Add(new Claim(Constants.ClaimTypes.GivenName, user[FIELD_USUARIO_NOMBRE]));
            }

**Comment by ahokkonen-dev** [comment]
I'm not sure you are still fighting with this issue or it is still relevant, but....

I had the same issue which I was tried to solve with many solution (token_hint, empty claim etc)
After all I noticed that in my modified "_loggoedOut" template I was accidentally removed the iframe -tag where logouturls are rendered :)

So, if you are using custom templates, make sure you have this iframe on you logedout.html -view:

`<iframe class="signout" ng-repeat="url in model.iFrameUrls" ng-src="{{url}}"></iframe>`

**Comment by venkatrv** [comment]
I am also facing this issue. but I did not modify any template.. 
so with this problem, if I have two MVC applications, and if I logout from one MVC application, what happens to other MVC application? will still maintain the authentication cookies? or cookies from both MVC applications gets cleaned up?

**Comment by Disciplemike** [comment]
Hey, I know this may be an old issue, but if you made the mistake I made, I hope this helps.
Make sure your SignoutCleanup is NOT AUTHORIZED! It seems pretty obvious, in hindsight, but it took me hours to figure this out.
I had my application adding a filter of Authorize attribute to everything. Hence, I didn't notice it on the controller method.
I just added [AllowAnonymous] to the SignoutCleanup method and blamo... everything works!

I hope this saves someone some time.
